README.md
# RBTray

RBTray is a small Windows program that runs in the background and allows almost
any window to be minimized to the system tray by:

- Right-clicking its minimize button
- Shift-right-clicking its title bar
- Using the Windows-Alt-Control-Down hotkey

Note that not all of these methods will work for every window, so please use
whichever one works for your needs.

RBTray is free, open source, and is distributed under the terms of the [GNU
General Public Licence](http://www.gnu.org/copyleft/gpl.html).

## Download

- [64 bit binaries](x64)
- [32 bit binaries](x86)
- [Original RBTray](https://sourceforge.net/projects/rbtray/files/)

## Installing

Download either the 32-bit or 64-bit binaries (depending on your OS) to a folder,
for example "`C:\Program Files\RBTray`".  Double-click RBTray.exe to start it. If
you want it to automatically start after you reboot, create a shortcut to
RBTray.exe in your Start menu's Startup group.

On Windows 10 that is located at
`"C:\Users\YOUR_USERNAME_HERE\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Startup"`

But a ***much*** quicker way to access it is to press "`WIN` + `R`" then type
(or copy & paste) `shell:Startup`

## Using

To minimize a program to the system tray, you can use any of these methods:

- Right-click with the mouse on the program's minimize button.
- Hold down a Shift key while right-clicking on the program's title bar.
- Pressing Windows-Alt-Control-Down on the keyboard (all at the same time).

This should create an icon for the window in the system tray. To restore the
program's window, single-click the program's icon in the tray. Alternatively,
you can right-click on the tray icon which will bring up a popup menu, then
select **Restore Window**.

In some cases, the first two methods cause problems with other software because
of they way they integrate into Windows using a hook to intercept mouse events.
In these cases, you can use the `--no-hook` option, which means that only the
last method of using the Windows-Alt-Control-Down hotkey will work. Also in
these cases the RBHook.dll isn't needed since it only exists to support the mouse
event hook.

## Exiting

Right-click on any tray icon created by RBTray and click **Exit RBTray** in the
popup menu, or run RBTray.exe with the `--exit` parameter.

## Authors

Nikolay Redko: http://rbtray.sourceforge.net/, https://github.com/nredko

J.D. Purcell: http://www.moitah.net/, https://github.com/jdpurcell

Benbuck Nason: https://github.com/benbuck

## Other

For original forum, bug tracker, etc. see [RBTray SourceForge project page]
(https://sourceforge.net/projects/rbtray/).

Copyright &copy; 1998-2011 Nikolay Redko, J.D. Purcell

Copyright &copy; 2015 Benbuck Nason


CHANGELOG.md
# Change Log

All notable changes to "RBTray" will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## Change history

## [4.14] fork (2022-Mar-20) - Benbuck Nason

- Change hotkey to Control-Alt-Down for simplicity and compatibility.
- Minor markdown cleanup/improvements.

## [4.13] fork (2021-Oct-09) - Benbuck Nason

- Change hotkey to Windows-Alt-Control-Down for Windows 10 and 11 compatibility.

## [4.12] fork (2021-Oct-07) - Benbuck Nason

- Change hotkey to Windows-Control-Down for Windows 11 compatibility (thanks to [Dimitri Pappas](https://github.com/fragtion)).
- Update project files to vstudio 2019

## [4.11] fork (2020-Mar-23) - Benbuck Nason

- Formatting improvements to RBTray.rc.
- Fix incorrect versions in RBTray.rc.
- Minor formatting improvements to README.md.
- Add --no-hook documentation to README.md.
- Added --no-hook command line option (thanks to [Ondrej Petrzilka](https://github.com/OndrejPetrzilka)).

## [4.10] fork (2019-Oct-03) - Benbuck Nason

- Fix missing tray icon issue (thanks to [Andrzej Włoszczyński](https://github.com/Andrzej-W)).

## [4.9] fork (2019-Oct-02) - Benbuck Nason

- Improved documentation.
- Clearer license/copyright notices (thanks to [Hugo Locurcio](https://github.com/Calinou)).
- Better error checking.
- Better command line parsing.

## [4.8] fork (2018-Jul-27) - Benbuck Nason

- Don't minimize desktop/taskbar/etc.
- Return correct value from WinMain.
- Allow processing of quit message.
- Clean up handling of WM_TASKBAR_CREATED.

## [4.7] fork (2018-Jul-21) - Benbuck Nason

- Support minimizing with Win-Alt-Down hotkey.

## [4.6] fork (2017-Aug-29) - Benbuck Nason

- Fix compatibility issue with some Windows 10 64 bit environments.

## [4.5] fork (2017-May-12) - Benbuck Nason

- Add some error handling.
- Fix bug in shift key check.
- Update project files to VS2015.

## [4.4] fork (2015-Dec-05) - Benbuck Nason

- Support minimizing by shift-right-clicking on window title bar.

## 4.3 (2011-Oct-30)

- Hide windows instead of minimizing and hiding.
- Hide windows even if they're already in the tray.
- Fix "Close Window" for Explorer windows.

## 4.2 (2011-Jun-24)

- Improve ability to minimize certain programs.

## 4.1 (2010-May-08)

- Unified mouse hook procedure improves reliability and fixes small minimize target with XP theming.
- Fixed tooltip overflow for long window captions.

## 4.0 (2010-May-06)

- Alternative mouse hook procedure for Windows Vista/7 Aero compatibility.
- Fixed 64-bit compatibility and created 64-bit compiles.
- Removed window menu additions (e.g. Minimize in tray, Always on top), hotkey, and keyboard indicator.
- Dropped Windows 9x support.
- Tray icon tooltip can display Unicode characters.
- Removes tray icon automatically if a program shows itself or exits.
- Fixed problem when minimizing Office 2007 windows.
- Won't minimize MDI child windows.
- Other small fixes and reorganized code.

[unreleased]: https://github.com/benbuck/rbtray/compare/v4.14...HEAD
[4.14]: https://github.com/benbuck/rbtray/compare/v4.13...v4.14
[4.13]: https://github.com/benbuck/rbtray/compare/v4.12...v4.13
[4.12]: https://github.com/benbuck/rbtray/compare/v4.11...v4.12
[4.11]: https://github.com/benbuck/rbtray/compare/v4.10...v4.11
[4.10]: https://github.com/benbuck/rbtray/compare/v4.9...v4.10
[4.9]: https://github.com/benbuck/rbtray/compare/v4.8...v4.9
[4.8]: https://github.com/benbuck/rbtray/compare/v4.7...v4.8
[4.7]: https://github.com/benbuck/rbtray/compare/v4.6...v4.7
[4.6]: https://github.com/benbuck/rbtray/compare/v4.5...v4.6
[4.5]: https://github.com/benbuck/rbtray/compare/v4.4...v4.5
[4.4]: https://github.com/benbuck/rbtray/releases/tag/v4.4


RBTray.h
// ****************************************************************************
//
// RBTray
// Copyright (C) 1998-2010  Nikolay Redko, J.D. Purcell
// Copyright (C) 2015 Benbuck Nason
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
// ****************************************************************************

#define NAME L"RBTrayHook"

#define WM_ADDTRAY      0x0401
#define WM_REMTRAY      0x0402
#define WM_REFRTRAY     0x0403
#define WM_TRAYCMD      0x0404
#define IDM_RESTORE     0x1001
#define IDM_CLOSE       0x1002
#define IDM_EXIT        0x1003
#define IDM_ABOUT       0x1004
#define IDM_RESTORE_ALL 0x1005

#define DLLIMPORT __declspec(dllexport)

BOOL DLLIMPORT RegisterHook(HMODULE);
void DLLIMPORT UnRegisterHook();


RBHook.cpp
// ****************************************************************************
//
// RBTray
// Copyright (C) 1998-2010  Nikolay Redko, J.D. Purcell
// Copyright (C) 2015 Benbuck Nason
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
// ****************************************************************************

#include <windows.h>
#include <stdio.h>
#include "RBTray.h"

static HHOOK _hMouse = NULL;
static HHOOK _hWndProcRet = NULL;
static HWND _hLastHit = NULL;


//#define DEBUG_PRINTF(fmt, ...) do { char buf[1024]; snprintf(buf, sizeof(buf), fmt, ##__VA_ARGS__); OutputDebugStringA(buf); } while (0)

// Works for 32-bit and 64-bit apps
LRESULT CALLBACK MouseProc(int nCode, WPARAM wParam, LPARAM lParam) {
    if (nCode >= 0) {
        //if ((wParam != WM_MOUSEMOVE) && (wParam != WM_NCMOUSEMOVE)) DEBUG_PRINTF("%s(%d): nCode %d, wParam %llx, lParam %llx\n", __FUNCTION__, __LINE__, nCode, (long long int)wParam, (long long int)lParam);
        MOUSEHOOKSTRUCT * info = (MOUSEHOOKSTRUCT *)lParam;
        //if ((wParam != WM_MOUSEMOVE) && (wParam != WM_NCMOUSEMOVE)) DEBUG_PRINTF("%s(%d): pt (%ld, %ld), hwnd %p, wHitTestCode %u, dwExtraInfo %llx\n", __FUNCTION__, __LINE__, info->pt.x, info->pt.y, info->hwnd, info->wHitTestCode, info->dwExtraInfo);
        if ((wParam == WM_NCRBUTTONDOWN) || (wParam == WM_NCRBUTTONUP)) {
            //DEBUG_PRINTF("%s(%d): button\n", __FUNCTION__, __LINE__);
            if (info->wHitTestCode == HTCLIENT) {
                //DEBUG_PRINTF("%s(%d): ignoring client hit test code for non-client message\n", __FUNCTION__, __LINE__);
            } else {
                BOOL shiftKeyDown = (GetKeyState(VK_SHIFT) & 0x8000) ? TRUE : FALSE;
                BOOL isHit = (info->wHitTestCode == HTMINBUTTON) || ((info->wHitTestCode == HTCAPTION) && shiftKeyDown);
                //DEBUG_PRINTF("%s(%d): shift %s, hit %s\n", __FUNCTION__, __LINE__, shiftKeyDown ? "yes" : "no", isHit ? "yes" : "no");
                if ((wParam == WM_NCRBUTTONDOWN) && isHit) {
                    //DEBUG_PRINTF("%s(%d): down hit\n", __FUNCTION__, __LINE__);
                    _hLastHit = info->hwnd;
                    return 1;
                } else if ((wParam == WM_NCRBUTTONUP) && isHit) {
                    //DEBUG_PRINTF("%s(%d): up hit\n", __FUNCTION__, __LINE__);
                    if (info->hwnd == _hLastHit) {
                        //DEBUG_PRINTF("%s(%d): up hit match\n", __FUNCTION__, __LINE__);
                        PostMessage(FindWindow(NAME, NAME), WM_ADDTRAY, 0, (LPARAM)info->hwnd);
                    }
                    _hLastHit = NULL;
                    return 1;
                } else {
                    //DEBUG_PRINTF("%s(%d): miss\n", __FUNCTION__, __LINE__);
                    _hLastHit = NULL;
                }
            }
        } else if ((wParam == WM_RBUTTONDOWN) || (wParam == WM_RBUTTONUP)) {
            //DEBUG_PRINTF("%s(%d): not button\n", __FUNCTION__, __LINE__);
            _hLastHit = NULL;
        }
    }
    return CallNextHookEx(_hMouse, nCode, wParam, lParam);
}

// Only works for 32-bit apps or 64-bit apps depending on whether this is complied
// as 32-bit or 64-bit
LRESULT CALLBACK CallWndProcRet(int nCode, WPARAM wParam, LPARAM lParam) {
    if (nCode >= 0) {
        CWPRETSTRUCT * msg = (CWPRETSTRUCT *)lParam;
        if (((msg->message == WM_WINDOWPOSCHANGED) &&
            (((WINDOWPOS *)msg->lParam)->flags & SWP_SHOWWINDOW) != 0) ||
            (msg->message == WM_NCDESTROY))
        {
            PostMessage(FindWindow(NAME, NAME), WM_REFRTRAY, 0, (LPARAM)msg->hwnd);
        }

        //if ((msg->message == WM_SYSCOMMAND)) {// && (msg->wParam == SC_MINIMIZE)) {
        //    DEBUG_PRINTF("%s(%d): minimize\n", __FUNCTION__, __LINE__);
        //    WCHAR text[256];
        //    GetWindowText(msg->hwnd, text, sizeof(text) / sizeof(text[0]));
        //    if (wcsstr(text, L"Notepad")) {
        //        //PostMessage(FindWindow(NAME, NAME), WM_ADDTRAY, 0, (LPARAM)msg->hwnd);
        //    }
        //}
    }
    return CallNextHookEx(_hWndProcRet, nCode, wParam, lParam);
}

BOOL DLLIMPORT RegisterHook(HMODULE hLib) {
    //DEBUG_PRINTF("%s(%d): register hook\n", __FUNCTION__, __LINE__);
    _hMouse = SetWindowsHookEx(WH_MOUSE, (HOOKPROC)MouseProc, hLib, 0);
    _hWndProcRet = SetWindowsHookEx(WH_CALLWNDPROCRET, (HOOKPROC)CallWndProcRet, hLib, 0);
    if ((_hMouse == NULL) || (_hWndProcRet == NULL)) {
        UnRegisterHook();
        return FALSE;
    }
    return TRUE;
}

void DLLIMPORT UnRegisterHook() {
    //DEBUG_PRINTF("%s(%d): unregister hook\n", __FUNCTION__, __LINE__);
    if (_hMouse) {
        UnhookWindowsHookEx(_hMouse);
        _hMouse = NULL;
    }
    if (_hWndProcRet) {
        UnhookWindowsHookEx(_hWndProcRet);
        _hWndProcRet = NULL;
    }
}


v3/rbtray.h
//resources
#define CM_CLOSE                        101
#define CM_EXIT                         102
#define CM_HIDE                         103
#define IDD_DIALOG1                     1001
#define ID_OK                           1002
#define IDC_EDIT                        1003

#if !RC_INVOKED

#define NAME "RBTrayHook"

//commands
#define WM_MYCMD     0x0401
#define WM_TRAYCMD   0x0402
#define WM_QUERYTRAY 0x0403
#define WM_DELTRAY   0x0404
#define WM_DBG		 0x0405
#define IDM_CLOSE       0x1001
#define IDM_EXIT        0x1002
#define IDM_RESTORE     0x1003
#define IDM_ALWAYS_TRAY 0x1004
#define IDM_ABOUT       0x1005
#define IDM_DESTROY     0x1006
#define IDM_TRAY        0x1007
#define IDM_HIDE        0x1008
#define IDM_ONTOP       0x1009
#define IDM_SEPARATOR   0x1010
#define IDM_SIZE		0x1011
#define IDM_LOTUSEDIT   0x1012
#define IDM_KEY			0x1013
#define IDM_KEYSTAT     0x1014
#define IDM_SHOWKEY     0x1015
#define DLLIMPORT __declspec(dllexport)

BOOL DLLIMPORT RegisterHook(HMODULE) ;
void DLLIMPORT UnRegisterHook(void);
void DLLIMPORT SetProcessId(DWORD);
//typedef struct tagMOUSEHOOKSTRUCT {
//    POINT   pt;
//    HWND    hwnd;
//    UINT    wHitTestCode;
//    DWORD   dwExtraInfo;
//} MOUSEHOOKSTRUCT, FAR *LPMOUSEHOOKSTRUCT, *PMOUSEHOOKSTRUCT;
#endif


RBTray.cpp
// ****************************************************************************
//
// RBTray
// Copyright (C) 1998-2010  Nikolay Redko, J.D. Purcell
// Copyright (C) 2015 Benbuck Nason
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
// ****************************************************************************

#include <windows.h>
#include "RBTray.h"
#include "resource.h"

#define MAXTRAYITEMS 64
#define HK_HIDEWINDOW 0
#define HK_SHOWWINDOW 1

static UINT WM_TASKBAR_CREATED;

static HINSTANCE _hInstance;
static HMODULE _hLib;
static HWND _hwndHook;
static HWND _hwndItems[MAXTRAYITEMS];
static HWND _hwndForMenu;

int FindInTray(HWND hwnd) {
    for (int i = 0; i < MAXTRAYITEMS; i++) {
        if (_hwndItems[i] == hwnd) {
            return i;
        }
    }
    return -1;
}

HICON GetWindowIcon(HWND hwnd) {
    HICON icon;
    if (icon = (HICON)SendMessage(hwnd, WM_GETICON, ICON_SMALL, 0)) {
        return icon;
    }
    if (icon = (HICON)SendMessage(hwnd, WM_GETICON, ICON_BIG, 0)) {
        return icon;
    }
    if (icon = (HICON)GetClassLongPtr(hwnd, GCLP_HICONSM)) {
        return icon;
    }
    if (icon = (HICON)GetClassLongPtr(hwnd, GCLP_HICON)) {
        return icon;
    }
    return LoadIcon(NULL, IDI_WINLOGO);
}

static bool AddToTray(int i) {
    NOTIFYICONDATA nid;
    ZeroMemory(&nid, sizeof(nid));
    nid.cbSize           = NOTIFYICONDATA_V2_SIZE;
    nid.hWnd             = _hwndHook;
    nid.uID              = (UINT)i;
    nid.uFlags           = NIF_MESSAGE | NIF_ICON | NIF_TIP;
    nid.uCallbackMessage = WM_TRAYCMD;
    nid.hIcon            = GetWindowIcon(_hwndItems[i]);
    GetWindowText(_hwndItems[i], nid.szTip, sizeof(nid.szTip) / sizeof(nid.szTip[0]));
    nid.uVersion         = NOTIFYICON_VERSION;
    if (!Shell_NotifyIcon(NIM_ADD, &nid)) {
        return false;
    }
    if (!Shell_NotifyIcon(NIM_SETVERSION, &nid)) {
        Shell_NotifyIcon(NIM_DELETE, &nid);
        return false;
    }
    return true;
}

static bool AddWindowToTray(HWND hwnd) {
    int i = FindInTray(NULL);
    if (i == -1) {
        return false;
    }
    _hwndItems[i] = hwnd;
    return AddToTray(i);
}

static void MinimizeWindowToTray(HWND hwnd) {
    // Don't minimize MDI child windows
    if ((UINT)GetWindowLongPtr(hwnd, GWL_EXSTYLE) & WS_EX_MDICHILD) {
        return;
    }

    // If hwnd is a child window, find parent window (e.g. minimize button in
    // Office 2007 (ribbon interface) is in a child window)
    if ((UINT)GetWindowLongPtr(hwnd, GWL_STYLE) & WS_CHILD) {
        hwnd = GetAncestor(hwnd, GA_ROOT);
    }

    // Hide window before AddWindowToTray call because sometimes RefreshWindowInTray
    // can be called from inside ShowWindow before program window is actually hidden
    // and as a result RemoveWindowFromTray is called which immediately removes just
    // added tray icon.
    ShowWindow(hwnd, SW_MINIMIZE);
    ShowWindow(hwnd, SW_HIDE);

    // Add icon to tray if it's not already there
    if (FindInTray(hwnd) == -1) {
        if (!AddWindowToTray(hwnd)) {
          // If there is something wrong with tray icon restore program window.
            ShowWindow(hwnd, SW_RESTORE);
            ShowWindow(hwnd, SW_SHOW);
          SetForegroundWindow(hwnd);
          return;
        }
    }
}

static bool RemoveFromTray(int i) {
    NOTIFYICONDATA nid;
    ZeroMemory(&nid, sizeof(nid));
    nid.cbSize = NOTIFYICONDATA_V2_SIZE;
    nid.hWnd   = _hwndHook;
    nid.uID    = (UINT)i;
    if (!Shell_NotifyIcon(NIM_DELETE, &nid)) {
        return false;
    }
    return true;
}

static bool RemoveWindowFromTray(HWND hwnd) {
    int i = FindInTray(hwnd);
    if (i == -1) {
        return false;
    }
    if (!RemoveFromTray(i)) {
        return false;
    }
    _hwndItems[i] = NULL;
    return true;
}

static void RestoreWindowFromTray(HWND hwnd) {
    ShowWindow(hwnd, SW_RESTORE);
    ShowWindow(hwnd, SW_SHOW);
    SetForegroundWindow(hwnd);
    RemoveWindowFromTray(hwnd);
}

static void CloseWindowFromTray(HWND hwnd) {
    // Use PostMessage to avoid blocking if the program brings up a dialog on exit.
    // Also, Explorer windows ignore WM_CLOSE messages from SendMessage.
    PostMessage(hwnd, WM_CLOSE, 0, 0);

    Sleep(50);
    if (IsWindow(hwnd)) {
        Sleep(50);
    }

    if (!IsWindow(hwnd)) {
        // Closed successfully
        RemoveWindowFromTray(hwnd);
    }
}

void RestoreAllWindowsFromTray() {
    for (int i = 0; i < MAXTRAYITEMS; i++) {
        if (_hwndItems[i]) {
            RestoreWindowFromTray(_hwndItems[i]);
        }
    }
}

void RefreshWindowInTray(HWND hwnd) {
    int i = FindInTray(hwnd);
    if (i == -1) {
        return;
    }
    if (!IsWindow(hwnd) || IsWindowVisible(hwnd)) {
        RemoveWindowFromTray(hwnd);
    } else {
        NOTIFYICONDATA nid;
        ZeroMemory(&nid, sizeof(nid));
        nid.cbSize = NOTIFYICONDATA_V2_SIZE;
        nid.hWnd   = _hwndHook;
        nid.uID    = (UINT)i;
        nid.uFlags = NIF_TIP;
        GetWindowText(hwnd, nid.szTip, sizeof(nid.szTip) / sizeof(nid.szTip[0]));
        Shell_NotifyIcon(NIM_MODIFY, &nid);
    }
}

void ExecuteMenu() {
    HMENU hMenu;
    POINT point;

    hMenu = CreatePopupMenu();
    if (!hMenu) {
        MessageBox(NULL, L"Error creating menu.", L"RBTray", MB_OK | MB_ICONERROR);
        return;
    }
    AppendMenu(hMenu, MF_STRING, IDM_ABOUT,       L"About RBTray");
    AppendMenu(hMenu, MF_STRING, IDM_EXIT,        L"Exit RBTray");
    AppendMenu(hMenu, MF_SEPARATOR, 0, NULL);     //--------------
    AppendMenu(hMenu, MF_STRING, IDM_CLOSE,       L"Close Window");
    AppendMenu(hMenu, MF_STRING, IDM_RESTORE,     L"Restore Window");
    AppendMenu(hMenu, MF_STRING, IDM_RESTORE_ALL, L"Restore All");

    GetCursorPos(&point);
    SetForegroundWindow(_hwndHook);

    TrackPopupMenu(hMenu, TPM_LEFTBUTTON | TPM_RIGHTBUTTON | TPM_RIGHTALIGN | TPM_BOTTOMALIGN, point.x, point.y, 0, _hwndHook, NULL);

    PostMessage(_hwndHook, WM_USER, 0, 0);
    DestroyMenu(hMenu);
}

BOOL CALLBACK AboutDlgProc(HWND hWnd, UINT Msg, WPARAM wParam, LPARAM lParam) {
    switch (Msg) {
        case WM_CLOSE:
            PostMessage(hWnd, WM_COMMAND, IDCANCEL, 0);
            break;
        case WM_COMMAND:
            switch (LOWORD(wParam)) {
                case IDOK:
                    EndDialog(hWnd, TRUE);
                    break;
                case IDCANCEL:
                    EndDialog(hWnd, FALSE);
                    break;
            }
            break;
        default:
            return FALSE;
    }
    return TRUE;
}

LRESULT CALLBACK HookWndProc(HWND hwnd, UINT msg, WPARAM wParam, LPARAM lParam) {
    switch (msg) {
        case WM_COMMAND:
            switch (LOWORD(wParam)) {
                case IDM_RESTORE_ALL:
                    RestoreAllWindowsFromTray();
                    break;
                case IDM_RESTORE:
                    RestoreWindowFromTray(_hwndForMenu);
                    break;
                case IDM_CLOSE:
                    CloseWindowFromTray(_hwndForMenu);
                    break;
                case IDM_ABOUT:
                    DialogBox(_hInstance, MAKEINTRESOURCE(IDD_ABOUT), _hwndHook, (DLGPROC)AboutDlgProc);
                    break;
                case IDM_EXIT:
                    SendMessage(_hwndHook, WM_DESTROY, 0, 0);
                    break;
            }
            break;
        case WM_ADDTRAY:
            MinimizeWindowToTray((HWND)lParam);
            break;
        case WM_REMTRAY:
            RestoreWindowFromTray((HWND)lParam);
            break;
        case WM_REFRTRAY:
            RefreshWindowInTray((HWND)lParam);
            break;
        case WM_TRAYCMD:
            switch ((UINT)lParam) {
                case NIN_SELECT:
                    RestoreWindowFromTray(_hwndItems[wParam]);
                    break;
                case WM_CONTEXTMENU:
                    _hwndForMenu = _hwndItems[wParam];
                    ExecuteMenu();
                    break;
                case WM_MOUSEMOVE:
                    RefreshWindowInTray(_hwndItems[wParam]);
                    break;
            }
            break;
        case WM_HOTKEY: {
            if (wParam == HK_HIDEWINDOW) {
                HWND fgWnd = GetForegroundWindow();
                if (!fgWnd)
                    break;

                LONG style = GetWindowLong(fgWnd, GWL_STYLE);
                if (!(style & WS_MINIMIZEBOX)) {
                    // skip, no minimize box
                    break;
                }

                MinimizeWindowToTray(fgWnd);
            }

            if (wParam == HK_SHOWWINDOW) {
                RestoreAllWindowsFromTray();
            }
            break;
        }
        case WM_DESTROY: {
            RestoreAllWindowsFromTray();

            if (_hLib)
            {
                UnRegisterHook();
                FreeLibrary(_hLib);
            }
            PostQuitMessage(0);
            break;
        }
        default:
            if (msg == WM_TASKBAR_CREATED) {
                for (int i = 0; i < MAXTRAYITEMS; i++) {
                    if (_hwndItems[i]) {
                        AddToTray(i);
                    }
                }
            }
            break;
    }

    return DefWindowProc(hwnd, msg, wParam, lParam);
}

int WINAPI WinMain(_In_ HINSTANCE hInstance, _In_opt_ HINSTANCE /*hPrevInstance*/, _In_ LPSTR /*szCmdLine*/, _In_ int /*iCmdShow*/) {
    _hInstance = hInstance;

    int argc;
    LPWSTR * argv = CommandLineToArgvW(GetCommandLineW(), &argc);
    bool shouldExit = false;
    bool useHook = true;
    for (int a = 0; a < argc; ++a) {
        if (!wcscmp(argv[a], L"--exit")) {
            shouldExit = true;
        }
        if (!wcscmp(argv[a], L"--no-hook")) {
            useHook = false;
        }
    }

    _hwndHook = FindWindow(NAME, NAME);
    if (_hwndHook) {
        if (shouldExit) {
            SendMessage(_hwndHook, WM_CLOSE, 0, 0);
        } else {
            MessageBox(NULL, L"RBTray is already running.", L"RBTray", MB_OK | MB_ICONINFORMATION);
        }
        return 0;
    }

    if (useHook) {
        if (!(_hLib = LoadLibrary(L"RBHook.dll"))) {
            MessageBox(NULL, L"Error loading RBHook.dll.", L"RBTray", MB_OK | MB_ICONERROR);
            return 0;
        }
        if (!RegisterHook(_hLib)) {
            MessageBox(NULL, L"Error setting hook procedure.", L"RBTray", MB_OK | MB_ICONERROR);
            return 0;
        }
    }

    WNDCLASS wc;
    wc.style         = 0;
    wc.lpfnWndProc   = HookWndProc;
    wc.cbClsExtra    = 0;
    wc.cbWndExtra    = 0;
    wc.hInstance     = hInstance;
    wc.hIcon         = NULL;
    wc.hCursor       = NULL;
    wc.hbrBackground = (HBRUSH)(COLOR_WINDOW + 1);
    wc.lpszMenuName  = NULL;
    wc.lpszClassName = NAME;
    if (!RegisterClass(&wc)) {
        MessageBox(NULL, L"Error creating window class", L"RBTray", MB_OK | MB_ICONERROR);
        return 0;
    }

    if (!(_hwndHook = CreateWindow(NAME, NAME, WS_OVERLAPPED, 0, 0, 0, 0, (HWND)NULL, (HMENU)NULL, (HINSTANCE)hInstance, (LPVOID)NULL))) {
        MessageBox(NULL, L"Error creating window", L"RBTray", MB_OK | MB_ICONERROR);
        return 0;
    }

    for (int i = 0; i < MAXTRAYITEMS; i++) {
        _hwndItems[i] = NULL;
    }

    WM_TASKBAR_CREATED = RegisterWindowMessage(L"TaskbarCreated");

    BOOL registeredACDHotKey = RegisterHotKey(_hwndHook, HK_HIDEWINDOW, MOD_ALT | MOD_CONTROL | MOD_NOREPEAT, VK_DOWN);
    if (!registeredACDHotKey) {
        MessageBox(NULL, L"Couldn't register ACD hotkey", L"RBTray", MB_OK | MB_ICONERROR);
    }

    BOOL registeredACUHotKey = RegisterHotKey(_hwndHook, HK_SHOWWINDOW, MOD_ALT | MOD_CONTROL | MOD_NOREPEAT, VK_UP);
    if (!registeredACUHotKey) {
        MessageBox(NULL, L"Couldn't register ACU hotkey", L"RBTray", MB_OK | MB_ICONERROR);
    }

    MSG msg;
    while (GetMessage(&msg, nullptr, 0, 0)) {
        TranslateMessage(&msg);
        DispatchMessage(&msg);
    }

    if (registeredACDHotKey) {
        UnregisterHotKey(_hwndHook, HK_HIDEWINDOW);
    }

    if (registeredACUHotKey) {
        UnregisterHotKey(_hwndHook, HK_SHOWWINDOW);
    }

    return (int)msg.wParam;
}
